File: argon-plugin/shared/src/main/scala/dev/argon/plugin/Platform.scala
package dev.argon.plugin

trait Platform[E] {
  val id: String
  val name: String

  type ExternFunction
  type ExternMethod
  type ExternClassConstructor

  val externFunctionCodec: ExternCodec[E, ExternFunction]
  val externMethodCodec: ExternCodec[E, ExternMethod]
  val externClassConstructorCodec: ExternCodec[E, ExternClassConstructor]
}

File: plugins/java/api/src/dev/argon/plugin/api/Platform.java
package dev.argon.plugin.api;

public interface Platform<E extends Exception, ExternFunction, ExternMethod, ExternClassCtor> {
  String id();
  String name();

  ExternCodec<E, ExternFunction> externFunctionCodec();
  ExternCodec<E, ExternMethod> externMethodCodec();
  ExternCodec<E, ExternClassCtor> externClassConstructorCodec();
}
